src/odemis/driver/actuator.py

```diff
@@ -940,6 +940,203 @@ def reference(self, axes):
+        positions (set or dict value -> str): positions where the actuator is allowed to move
+        cycle (float): if not None, it means the actuator does a cyclic move and this value represents a full cycle
+        """
+        if inverted:
+            raise ValueError("Axes shouldn't be inverted")
+
+        if len(children) != 1:
+            raise ValueError("FixedPositionsActuator needs precisely one child")
+
+        self._move_sum = 0
+        self._position = {}
+        self._referenced = {}
+        axis, child = children.items()[0]
+        self._axis = axis
+        self._child = child
+        self._caxis = axis_name
+        self._offset = offset
+
+        self._moves_before_ref = moves_before_ref  # max number of moves before referencing
+        self._move_num_total = 0  # current number of moves after last referencing
+
+        # Executor used to reference and move to nearest position
+        self._executor = CancellableThreadPoolExecutor(max_workers=1)  # one task at a time
+
+        if not isinstance(child, model.ComponentBase):
+            raise ValueError("Child %s is not a component." % (child,))
+        if not hasattr(child, "axes") or not isinstance(child.axes, dict):
+            raise ValueError("Child %s is not an actuator." % child.name)
+
+        ac = child.axes[axis_name]
+        rng = numpy.subtract(ac.range, [self._offset, self._offset])
+        axes = {axis: model.Axis(range=rng, unit=ac.unit)}  # TODO: allow the user to override the unit?
+
+        model.Actuator.__init__(self, name, role, axes=axes, children=children, **kwargs)
+
+        self._position = {}
+        self.position = model.VigilantAttribute({}, readonly=True)
+
+        logging.debug("Subscribing to position of child %s", child.name)
+        child.position.subscribe(self._update_child_position, init=True)
+
+        if model.hasVA(child, "referenced") and axis_name in child.referenced.value:
+            self._referenced[axis] = child.referenced.value[axis_name]
+            self.referenced = model.VigilantAttribute(self._referenced.copy(), readonly=True)
+            child.referenced.subscribe(self._update_child_ref)
+
+        # If the axis can be referenced => do it now (and move to a known position)
+        # In case of cyclic move always reference
+        if not self._referenced.get(axis, True):
+            # The initialisation will not fail if the referencing fails
+            f = self.reference({axis})
+            f.add_done_callback(self._on_referenced)
+        else:
+            self.moveAbs({self._axis: self._child.position.value[self._caxis]}).result()
+
+    def _on_referenced(self, future):
+        try:
+            future.result()
+        except Exception as e:
+            self._child.stop({self._caxis})  # prevent any move queued
+            self.state._set_value(e, force_write=True)
+            logging.exception(e)
+
+    def _update_child_position(self, value):
+        p = value[self._caxis]
```

Comment:
This is quite convoluted to store a single value. Can you maybe just write `self._pos = value[self._caxis] - self._offset`, and be done?